File: code/apps/web/src/lib/workspace/navigation.ts
/**
 * Role-specific workspace home (SRS-WEB-004).
 *
 * The verification clause is the whole design: "navigation reflects role but
 * server still enforces authorization". Those are two independent statements,
 * and conflating them is the classic front-end security failure — a menu that
 * hides an action is a menu, not a control, and the endpoint behind it is
 * still there for anyone who knows the URL.
 *
 * So this module builds navigation from permissions and says so loudly, and a
 * test asserts that nothing here is used as an access decision. Hiding is a
 * courtesy: it stops the user being offered buttons that will fail. The server
 * refuses regardless.
 */

/** One entry in the workspace navigation. */
export interface NavItem {
	readonly id: string;
	readonly label: string;
	readonly href: string;
	/**
	 * The permission that makes this entry useful. The server enforces it; this
	 * only decides whether to offer the link.
	 */
	readonly requires: string;
	/** Grouping in the sidebar. */
	readonly section: 'clinical' | 'operations' | 'administration';
}

/** A quick action offered on the workspace home. */
export interface QuickAction {
	readonly id: string;
	readonly label: string;
	readonly requires: string;
	/**
	 * True for actions that finalize something clinical or financial. The home
	 * screen renders these differently — never as a one-click tile — because a
	 * quick action that signs something is a mis-click away from a clinical
	 * event (SRS-WEB-007).
	 */
	readonly finalizes: boolean;
}

/** What a worklist tile shows. */
export interface Worklist {
	readonly id: string;
	readonly label: string;
	readonly requires: string;
	/** Route the tile links to. */
	readonly href: string;
}

/** The catalogue the workspace is built from. */
export interface WorkspaceCatalogue {
	readonly navigation: readonly NavItem[];
	readonly quickActions: readonly QuickAction[];
	readonly worklists: readonly Worklist[];
}

/** What a user actually sees. */
export interface Workspace {
	readonly navigation: readonly NavItem[];
	readonly quickActions: readonly QuickAction[];
	readonly worklists: readonly Worklist[];
	/**
	 * True when the user holds no permission that any entry needs. The home
	 * screen shows an explanation rather than an empty page, because an empty
	 * workspace looks identical to a broken one.
	 */
	readonly empty: boolean;
}

/**
 * Builds the workspace a set of permissions should see.
 *
 * Pure and permission-driven. It takes the permission list rather than a
 * session object so it cannot accidentally reach for anything else — a
 * function that could see the whole session could start branching on a role
 * name, and role names drift from permissions the moment a role is renamed.
 */
export function buildWorkspace(
	catalogue: WorkspaceCatalogue,
	permissions: readonly string[]
): Workspace {
	const held = new Set(permissions);
	const offers = (requires: string): boolean => held.has(requires);

	const navigation = catalogue.navigation.filter((i) => offers(i.requires));
	const quickActions = catalogue.quickActions.filter((a) => offers(a.requires));
	const worklists = catalogue.worklists.filter((w) => offers(w.requires));

	return {
		navigation,
		quickActions,
		worklists,
		empty: navigation.length === 0 && quickActions.length === 0 && worklists.length === 0
	};
}

/**
 * The Wave-0 catalogue.
 *
 * Small, because Wave 0 has two bounded contexts. It is a data structure
 * rather than markup so a test can assert what each role sees without
 * rendering anything, and so a Wave-1 context adds entries without touching a
 * component.
 */
export const waveZeroCatalogue: WorkspaceCatalogue = {
	navigation: [
		{
			id: 'facilities',
			label: 'Facilities',
			href: '/facilities',
			requires: 'organization.facility.read',
			section: 'administration'
		},
		{
			id: 'org-units',
			label: 'Departments and units',
			href: '/units',
			requires: 'organization.unit.read',
			section: 'administration'
		},
		{
			id: 'approvals',
			label: 'Pending approvals',
			href: '/approvals',
			requires: 'organization.master_data.approve',
			section: 'operations'
		}
	],
	quickActions: [
		{
			id: 'new-facility',
			label: 'Add facility',
			requires: 'organization.facility.create',
			finalizes: false
		},
		{
			id: 'approve-change',
			label: 'Approve a master-data change',
			requires: 'organization.master_data.approve',
			// Approving a change takes effect on production configuration, so
			// it is never a one-click tile.
			finalizes: true
		}
	],
	worklists: [
		{
			id: 'pending-approvals',
			label: 'Changes awaiting your approval',
			requires: 'organization.master_data.approve',
			href: '/approvals'
		},
		{
			id: 'emergency-review',
			label: 'Emergency access awaiting review',
			requires: 'security.emergency.review',
			href: '/security/emergency'
		}
	]
};

/**
 * The Wave-1 catalogue: the reception desk (UX-W1-01).
 *
 * Separate from the Wave-0 entries rather than merged into them, so a
 * deployment running only the platform contexts is not offered clinical
 * navigation that would fail at the server. Merged at the composition point
 * below.
 */
export const receptionCatalogue: WorkspaceCatalogue = {
	navigation: [
		{
			id: 'reception-board',
			label: 'Reception board',
			href: '/reception',
			requires: 'sch.schedule.read',
			section: 'operations'
		},
		{
			id: 'patient-search',
			label: 'Find a patient',
			href: '/reception/search',
			requires: 'empi.patient.read',
			section: 'clinical'
		},
		{
			id: 'ward-worklist',
			label: 'Ward worklist',
			href: '/ward',
			requires: 'nursing.record.read',
			section: 'clinical'
		},
		{
			id: 'orders',
			label: 'Orders and results',
			href: '/orders',
			requires: 'orders.order.read',
			section: 'clinical'
		}
	],
	quickActions: [
		{
			id: 'find-patient',
			label: 'Find a patient',
			requires: 'empi.patient.read',
			finalizes: false
		},
		{
			id: 'register-patient',
			label: 'Register a patient',
			// Creating a patient record is not reversible by the person who did
			// it, and a duplicate created here is read as a second person for as
			// long as it takes somebody to notice. Never a one-click tile: the
			// action routes into search-before-create (SRS-EMPI-003).
			requires: 'empi.patient.create',
			finalizes: true
		}
	],
	worklists: [
		{
			id: 'todays-clinic',
			label: "Today's clinic",
			requires: 'sch.schedule.read',
			href: '/reception'
		},
		{
			id: 'nursing-work',
			label: 'Nursing work and reassessments',
			requires: 'nursing.record.read',
			href: '/ward'
		},
		{
			id: 'results-inbox',
			label: 'Critical results awaiting action',
			requires: 'clinical.record.write',
			href: '/orders'
		}
	]
};

/** Merges catalogues into the one the shell renders. */
export function mergeCatalogues(
	...catalogues: readonly WorkspaceCatalogue[]
): WorkspaceCatalogue {
	return {
		navigation: catalogues.flatMap((c) => c.navigation),
		quickActions: catalogues.flatMap((c) => c.quickActions),
		worklists: catalogues.flatMap((c) => c.worklists)
	};
}

File: code/apps/mobile/lib/src/workspace/navigation.dart
/// Role-specific workspace home (SRS-WEB-004).
///
/// The verification clause is "navigation reflects role but server still
/// enforces authorization". Those are two independent statements, and
/// conflating them is the classic client-side security failure: a menu that
/// hides an action is a menu, not a control, and the endpoint behind it is
/// still there for anybody who can reach it.
///
/// So this builds navigation from permissions and says so loudly, and a test
/// asserts that nothing here is used as an access decision. Hiding is a
/// courtesy: it stops the user being offered actions that will fail. The
/// server refuses regardless.
library;

import 'package:meta/meta.dart';

/// Where an entry sits in the app.
enum WorkspaceSection { clinical, operations, administration }

/// One entry in the workspace navigation.
@immutable
class NavItem {
  const NavItem({
    required this.id,
    required this.label,
    required this.route,
    required this.requires,
    required this.section,
  });

  final String id;
  final String label;
  final String route;

  /// The permission that makes this entry useful. The server enforces it; this
  /// only decides whether to offer the link.
  final String requires;

  final WorkspaceSection section;
}

/// A quick action offered on the workspace home.
@immutable
class QuickAction {
  const QuickAction({
    required this.id,
    required this.label,
    required this.requires,
    required this.finalizes,
  });

  final String id;
  final String label;
  final String requires;

  /// True for actions that finalize something clinical or financial. The home
  /// screen renders these differently — never as a one-tap tile, because a tap
  /// target on a phone is hit by accident far more often than a mouse click.
  final bool finalizes;
}

/// What a worklist tile shows.
@immutable
class Worklist {
  const Worklist({
    required this.id,
    required this.label,
    required this.requires,
    required this.route,
  });

  final String id;
  final String label;
  final String requires;
  final String route;
}

/// The catalogue the workspace is built from.
@immutable
class WorkspaceCatalogue {
  const WorkspaceCatalogue({
    required this.navigation,
    required this.quickActions,
    required this.worklists,
  });

  final List<NavItem> navigation;
  final List<QuickAction> quickActions;
  final List<Worklist> worklists;
}

/// What a user actually sees.
@immutable
class Workspace {
  const Workspace({
    required this.navigation,
    required this.quickActions,
    required this.worklists,
    required this.empty,
  });

  final List<NavItem> navigation;
  final List<QuickAction> quickActions;
  final List<Worklist> worklists;

  /// True when the user holds no permission that any entry needs. The home
  /// screen shows an explanation rather than a blank page, because an empty
  /// workspace looks identical to a broken one.
  final bool empty;
}

/// Builds the workspace a set of permissions should see.
///
/// Takes the permission list rather than a session object so it cannot
/// accidentally reach for anything else — a function that could see the whole
/// session could start branching on a role name, and role names drift from
/// permissions the moment a role is renamed.
Workspace buildWorkspace(
  WorkspaceCatalogue catalogue,
  List<String> permissions,
) {
  final held = permissions.toSet();
  bool offers(String requires) => held.contains(requires);

  final navigation =
      catalogue.navigation.where((i) => offers(i.requires)).toList(growable: false);
  final quickActions =
      catalogue.quickActions.where((a) => offers(a.requires)).toList(growable: false);
  final worklists =
      catalogue.worklists.where((w) => offers(w.requires)).toList(growable: false);

  return Workspace(
    navigation: navigation,
    quickActions: quickActions,
    worklists: worklists,
    empty: navigation.isEmpty && quickActions.isEmpty && worklists.isEmpty,
  );
}

/// The Wave-0 catalogue.
///
/// Small, because Wave 0 has two bounded contexts. A data structure rather than
/// widgets so a test can assert what each role sees without pumping anything,
/// and so a Wave-1 context adds entries without touching a screen.
const WorkspaceCatalogue waveZeroCatalogue = WorkspaceCatalogue(
  navigation: [
    NavItem(
      id: 'facilities',
      label: 'Facilities',
      route: '/facilities',
      requires: 'organization.facility.read',
      section: WorkspaceSection.administration,
    ),
    NavItem(
      id: 'org-units',
      label: 'Departments and units',
      route: '/units',
      requires: 'organization.unit.read',
      section: WorkspaceSection.administration,
    ),
  ],
  quickActions: [
    QuickAction(
      id: 'new-facility',
      label: 'Add facility',
      requires: 'organization.facility.create',
      finalizes: false,
    ),
    QuickAction(
      id: 'approve-change',
      label: 'Approve a master-data change',
      requires: 'organization.master_data.approve',
      // Approving a change takes effect on production configuration, so it is
      // never a one-tap tile.
      finalizes: true,
    ),
  ],
  worklists: [
    Worklist(
      id: 'pending-approvals',
      label: 'Changes awaiting your approval',
      requires: 'organization.master_data.approve',
      route: '/approvals',
    ),
  ],
);

/// The Wave-1 catalogue for a ward device.
///
/// Permission-gated rather than short. An earlier version of this file carried
/// only the two workspaces a nurse uses on a round, on the argument that a
/// tablet is for work done standing up. That argument was about the device and
/// it was the wrong axis: the same tablet is carried by a receptionist at a
/// counter and handed to a clinician at a bedside, and what each of them
/// should see is decided by what they may do, not by what the hardware is.
///
/// So every workspace is here, and each appears only for somebody holding the
/// permission it needs. A nurse still sees a short list — theirs — because the
/// permissions do the narrowing that the old comment did by omission.
const WorkspaceCatalogue wardCatalogue = WorkspaceCatalogue(
  navigation: [
    NavItem(
      id: 'reception',
      label: 'Reception',
      route: '/reception',
      requires: 'scheduling.queue.read',
      section: WorkspaceSection.clinical,
    ),
    NavItem(
      id: 'patients',
      label: 'My patients',
      route: '/patients',
      // The same permission the worklist needs: a nurse who may see the work
      // may see whose work it is.
      requires: 'nursing.task.read',
      section: WorkspaceSection.clinical,
    ),
    NavItem(
      id: 'chart',
      label: 'Chart',
      route: '/chart',
      requires: 'clinical.note.read',
      section: WorkspaceSection.clinical,
    ),
    NavItem(
      id: 'ward',
      label: 'Ward worklist',
      route: '/ward',
      requires: 'nursing.task.read',
      section: WorkspaceSection.clinical,
    ),
    NavItem(
      id: 'medication-round',
      label: 'Medication round',
      route: '/medications/round',
      requires: 'nursing.administration.read',
      section: WorkspaceSection.clinical,
    ),
  ],
  quickActions: [
    QuickAction(
      id: 'chart-observation',
      label: 'Record observations',
      requires: 'nursing.observation.write',
      finalizes: false,
    ),
    QuickAction(
      id: 'administer',
      label: 'Give a medication',
      // Finalizing: this one puts a drug into a patient, and on a phone a tile
      // is hit by accident far more often than a mouse click.
      requires: 'nursing.administration.write',
      finalizes: true,
    ),
  ],
  worklists: [
    Worklist(
      id: 'doses-due',
      label: 'Doses due now',
      requires: 'nursing.administration.read',
      route: '/medications/round',
    ),
    Worklist(
      id: 'observations-due',
      label: 'Observations due',
      requires: 'nursing.task.read',
      route: '/ward',
    ),
  ],
);

/// Merges catalogues into the one the shell renders.
WorkspaceCatalogue mergeCatalogues(List<WorkspaceCatalogue> catalogues) =>
    WorkspaceCatalogue(
      navigation: [for (final c in catalogues) ...c.navigation],
      quickActions: [for (final c in catalogues) ...c.quickActions],
      worklists: [for (final c in catalogues) ...c.worklists],
    );
